# Assignment: Hypothesis Testing with Z-Test and t-Test
### Left-tailed, right-tailed and two-tailed tests on a student dataset

**Total marks:** 100  **Submit:** one `.ipynb` file

**What you will do**

You are given a small dataset of **40 college students**. Answer **6 questions**:

| Question | Test | Tail |
|---|---|---|
| Q1 | Z-test | Left |
| Q2 | Z-test | Right |
| Q3 | Z-test | Two |
| Q4 | t-test | Left |
| Q5 | t-test | Right |
| Q6 | t-test | Two |

Then fill in the summary table in Q7.

**Instructions**
1. Run the two starter cells first. **Do not change them.**
2. Write your code only in the cells marked `YOUR CODE HERE`.
3. Write your answers in the cells marked **Your answer**.
4. Use **α = 0.05** for every question.
5. Allowed libraries: `numpy`, `pandas`, `scipy.stats`, `matplotlib`.
6. Compute every number from the dataset. Do not type in the answers by hand.

## Quick Reference

| | Z-test | t-test |
|---|---|---|
| Use when | Population σ is **known** | Population σ is **unknown** |
| Formula | $z = \dfrac{\bar{x} - \mu_0}{\sigma / \sqrt{n}}$ | $t = \dfrac{\bar{x} - \mu_0}{s / \sqrt{n}}$, $df = n - 1$ |

| Tail | $H_1$ | p-value (Z) | p-value (t) | Critical value (Z, α = 0.05) |
|---|---|---|---|---|
| Left | $\mu < \mu_0$ | `stats.norm.cdf(z)` | `stats.t.cdf(t, dof)` | −1.645 |
| Right | $\mu > \mu_0$ | `1 - stats.norm.cdf(z)` | `1 - stats.t.cdf(t, dof)` | +1.645 |
| Two | $\mu \neq \mu_0$ | `2 * (1 - stats.norm.cdf(abs(z)))` | `2 * (1 - stats.t.cdf(abs(t), dof))` | ±1.960 |

**Decision rule:** if **p-value ≤ α**, reject $H_0$. Otherwise, fail to reject $H_0$.

For the sample standard deviation use `df['column'].std()`. pandas uses `ddof=1` by default, which is what the t-test needs.

⚠️ The dataset is stored in a variable called `df`. Store the degrees of freedom in a variable called **`dof`**, not `df`, or you will overwrite the dataset.

## Setup and Dataset

In [1]:
# ---------------- STARTER CODE – DO NOT MODIFY ----------------
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

alpha = 0.05
print("Libraries loaded. alpha =", alpha)

Libraries loaded. alpha = 0.05


In [2]:
# ---------------- STARTER CODE – DO NOT MODIFY ----------------
# Dataset: 40 students from three sections of a college course
df = pd.DataFrame({
    'student_id':  [f'S{i:02d}' for i in range(1, 41)],
    'section':     ['A'] * 14 + ['B'] * 13 + ['C'] * 13,
    'sleep_hours': [6.9, 5.5, 7.4, 7.6, 4.5, 5.2, 6.7, 6.3, 6.6, 5.7, 7.6, 7.5, 6.7, 7.8,
                    7.1, 5.7, 7.0, 5.5, 7.6, 6.5, 6.4, 5.9, 7.9, 6.4, 6.1, 6.2, 7.2,
                    7.0, 7.1, 7.1, 9.0, 6.2, 6.0, 5.7, 7.3, 7.8, 6.5, 5.7, 5.7, 7.3],
    'screen_time': [6.5, 6.2, 4.5, 5.8, 5.6, 5.8, 6.7, 5.8, 6.4, 5.5, 5.9, 6.3, 3.4, 5.0,
                    4.8, 4.6, 5.1, 7.5, 4.2, 6.8, 3.1, 5.0, 5.7, 6.3, 6.4, 6.6, 5.0,
                    4.8, 6.7, 5.2, 3.7, 3.9, 4.2, 6.1, 5.6, 6.4, 4.9, 5.7, 6.3, 5.0],
    'exam_score':  [71, 60, 63, 63, 55, 71, 62, 67, 71, 71, 73, 66, 62, 66,
                    50, 52, 53, 57, 70, 57, 63, 79, 63, 74, 57, 64, 57,
                    63, 75, 49, 71, 69, 61, 52, 67, 61, 69, 67, 83, 64],
    'attendance':  [75, 82, 82, 89, 86, 83, 90, 74, 77, 75, 79, 73, 85, 80,
                    72, 75, 83, 86, 93, 98, 83, 75, 68, 83, 76, 79, 77,
                    80, 87, 82, 80, 75, 71, 78, 81, 92, 82, 87, 78, 74],
    'study_hours': [2.2, 2.4, 4.7, 2.3, 3.7, 2.3, 3.7, 3.3, 2.9, 3.0, 2.5, 3.4, 2.6, 2.0,
                    2.0, 3.1, 4.3, 3.1, 2.9, 3.2, 4.0, 3.2, 2.7, 3.9, 3.3, 4.2, 3.1,
                    2.0, 1.9, 4.3, 4.4, 2.9, 2.7, 4.2, 2.1, 2.3, 3.5, 2.7, 3.0, 2.9],
    'commute_min': [34, 42, 33, 37, 18, 32, 26, 23, 26, 30, 38, 23, 32, 29,
                    30, 39, 36, 41, 31, 27, 30, 34, 33, 24, 33, 34, 50,
                    45, 26, 30, 22, 28, 34, 40, 27, 27, 17, 31, 25, 28],
})

print("Shape:", df.shape)
print(df['section'].value_counts().sort_index().to_string())
df.sample(10)

Shape: (40, 8)
section
A    14
B    13
C    13


,student_id,section,sleep_hours,screen_time,exam_score,attendance,study_hours,commute_min
17,S18,B,5.5,7.5,57,86,3.1,41
3,S04,A,7.6,5.8,63,89,2.3,37
14,S15,B,7.1,4.8,50,72,2.0,30
37,S38,C,5.7,5.7,67,87,2.7,31
23,S24,B,6.4,6.3,74,83,3.9,24
5,S06,A,5.2,5.8,71,83,2.3,32
8,S09,A,6.6,6.4,71,77,2.9,26
9,S10,A,5.7,5.5,71,75,3.0,30
31,S32,C,6.2,3.9,69,75,2.9,28
10,S11,A,7.6,5.9,73,79,2.5,38


**Data dictionary**

| Column | Meaning |
|---|---|
| `student_id` | Student ID |
| `section` | Class section: A (14 students), B (13), C (13) |
| `sleep_hours` | Average sleep per night (hours) |
| `screen_time` | Average daily screen time (hours) |
| `exam_score` | Final exam score (out of 100) |
| `attendance` | Attendance (%) |
| `study_hours` | Average self-study per day (hours) |
| `commute_min` | One-way commute time (minutes) |

Explore the data before you start (optional):

In [3]:
# ===================== OPTIONAL – EXPLORE THE DATA =====================
# OPTIONAL: explore the dataset, e.g. df.describe() or df.groupby('section').mean(numeric_only=True)


df.describe()

,sleep_hours,screen_time,exam_score,attendance,study_hours,commute_min
count,40.000000,40.00000,40.000000,40.000000,40.000000,40.000000
mean,6.647500,5.47500,64.200000,80.625000,3.072500,31.125000
std,0.900993,1.01571,7.855098,6.463656,0.761236,6.962123
min,4.500000,3.10000,49.000000,68.000000,1.900000,17.000000
25%,5.975000,4.87500,59.250000,75.000000,2.475000,26.750000
50%,6.650000,5.65000,63.500000,80.000000,3.000000,30.500000
75%,7.300000,6.30000,70.250000,83.500000,3.550000,34.000000
max,9.000000,7.50000,83.000000,98.000000,4.700000,50.000000


---
# Part A – Z-Tests (σ known, use all 40 students)

## Q1. Z-Test – Left-Tailed

A health report says college students sleep **7 hours** per night on average, with a known population standard deviation of **σ = 1.2 hours**.
Using `sleep_hours` for **all 40 students**, test at α = 0.05 whether students sleep **less than 7 hours** on average.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [4]:
# ===================== Q1 – YOUR CODE HERE =====================
# Z-test, left-tailed: column 'sleep_hours', mu_0 = 7, sigma = 1.2, all 40 students


sleep = np.array(df["sleep_hours"])

n = 40
x_bar = sleep.sum()/n
Mu = 7
sigma = 1.2

z = (x_bar - Mu)/(sigma/np.sqrt(n))

print("Z-value:- ", z)

p_value = stats.norm.cdf(z)

print("p-value:- " , p_value)

Z-value:-  -1.8578381253489278
p-value:-  0.03159600261338476


**Your answer – Q1**

- **$H_0$:college students sleep 7 hours on average**
- **$H_1$:college students sleep less than 7 hours on average**
- **Tail (left / right / two):left**
- **Test statistic:the z value is -1.857**
- **Critical value:-1.645**
- **p-value:0.0315**
- **Decision (Reject $H_0$ / Fail to reject $H_0$): Reject $H_0$**
- **Conclusion in words:p_value <= 0.05 means that the we have reject the H0 statment and conclude that  the college students on an average sleep less than 7 hours which also says that the college claims were false**

## Q2. Z-Test – Right-Tailed

A national survey says young adults spend **5 hours** a day on screens, with a known population standard deviation of **σ = 1.5 hours**.
Using `screen_time` for **all 40 students**, test at α = 0.05 whether these students' screen time is **more than 5 hours** a day.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [5]:
# ===================== Q2 – YOUR CODE HERE =====================
# Z-test, right-tailed: column 'screen_time', mu_0 = 5, sigma = 1.5, all 40 students

screen = np.array(df["screen_time"])

n = 40
x_bar = screen.sum()/n
Mu = 5
sigma = 1.5

z = (x_bar - Mu)/(sigma/np.sqrt(n))

print("Z-value:- ", z)

p_value = stats.norm.cdf(z)

print("p-value:- " , p_value)

Z-value:-  2.002775851439972
p-value:-  0.9773993235117389


**Your answer – Q2**

- **$H_0$:Adults on average spend 5 hours a day on the screen**
- **$H_1$:Adults on average spend more than 5 houts a day on the screen**
- **Tail (left / right / two):Right**
- **Test statistic:2.002**
- **Critical value:1.645**
- **p-value:0.977**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):Fail to reject $H_0$**
- **Conclusion in words:since the p value is more than 0.05 we can say that the claim made by the national survey is correct and average screen time of adults really is 5 hours a day**

## Q3. Z-Test – Two-Tailed

Over many years, the course's final exam has had a mean of **65** with a known standard deviation of **σ = 10**.
Using `exam_score` for **all 40 students**, test at α = 0.05 whether this year's mean score is **different from 65**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) find $n$ and $\bar{x}$, (c) compute $z$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion.

*(15 marks)*

In [6]:
# ===================== Q3 – YOUR CODE HERE =====================
# Z-test, two-tailed: column 'exam_score', mu_0 = 65, sigma = 10, all 40 students

score = np.array(df["exam_score"])

n = 40
x_bar = score.sum()/n
Mu = 65
sigma = 10

z = (x_bar - Mu)/(sigma/np.sqrt(n))

print("Z-value:- ", z)

p_value = stats.norm.cdf(z)

print("p-value:- " , p_value)

Z-value:-  -0.5059644256269389
p-value:-  0.30644081170661885


**Your answer – Q3**

- **$H_0$:The score of the final exam has mean of 65**
- **$H_1$:The score of the final exam is not 65**
- **Tail (left / right / two):two**
- **Test statistic:-0.505**
- **Critical value:+-1.96**
- **p-value:0.306**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):Fail to reject $H_0$**
- **Conclusion in words:The p value being more than 0.05 means that this year also the mean score of all the student's final exam is 65 and has not change over the past year**

---
# Part B – t-Tests (σ unknown, use one section only)

Select one section like this: `df[df['section'] == 'B']['attendance']`

## Q4. t-Test – Left-Tailed

The class teacher of **Section B** believes her students' attendance is **below 83 %**. The population standard deviation is **not known**.
Using `attendance` for **Section B only**, test at α = 0.05 whether the mean attendance is **less than 83 %**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [7]:
# ===================== Q4 – YOUR CODE HERE =====================
# t-test, left-tailed: column 'attendance', Section B only, mu_0 = 83, sigma unknown

attendance_B = np.array(df[df["section"] == "B"]["attendance"])

x_bar = attendance_B.mean()
Mu = 83
s = attendance_B.std(ddof =1)
n = len(attendance_B)
dof = n-1
t = (x_bar - Mu)/((s/np.sqrt(n)))
p_value = stats.t.cdf(t , dof)

print("t-test:- ", t)
print("p-value:- ", p_value)


t-test:-  -1.0306646945546007
p-value:-  0.16151225409771375


**Your answer – Q4**

- **$H_0$:The teacher clains the mean attendance of her students is less than 83**
- **$H_1$:The attendance is equal to 83 percentage**
- **Tail (left / right / two):left-tailed**
- **Test statistic:-1.03**
- **Critical value:-1.782**
- **p-value:0.1615**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):Fail to reject $H_0$**
- **Conclusion in words:the p value indicates that the mean attendance is 83 percentage not less than it**

## Q5. t-Test – Right-Tailed

The college recommends at least **2.5 hours** of self-study per day. The teacher of **Section A** claims her students study **more than 2.5 hours** a day on average. σ is **not known**.
Using `study_hours` for **Section A only**, test the claim at α = 0.05.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [9]:
# ===================== Q5 – YOUR CODE HERE =====================
# t-test, right-tailed: column 'study_hours', Section A only, mu_0 = 2.5, sigma unknown

study_A = np.array(df[df["section"] == "A"]["study_hours"])

x_bar = study_A.mean()
Mu = 2.5
s = study_A.std(ddof =1)
n = len(study_A)
dof = n-1
t = (x_bar - Mu)/((s/np.sqrt(n)))
p_value = 1 - stats.t.cdf(t , dof)

print("t-test:- ", t)
print("p-value:- ", p_value)

t-test:-  2.1184704732766493
p-value:-  0.02698523465440561


**Your answer – Q5**

- **$H_0$:The mean self-study time of Section A students is 2.5 hours**
- **$H_1$:The mean self-study time of Section A students is 2.5 hours**
- **Tail (left / right / two):Right**
- **Test statistic:2.118**
- **Critical value:1.771**
- **p-value:0.0269**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):Reject $H_0$**
- **Conclusion in words:Since the p-value is less than 0.05, we reject $H_0$. This means the average self-study time of Section A students is significantly more than 2.5 hours per day.**

## Q6. t-Test – Two-Tailed

The transport office plans the college bus timetable assuming an average one-way commute of **35 minutes**. σ is **not known**.
Using `commute_min` for **Section C only**, test at α = 0.05 whether the mean commute time is **different from 35 minutes**.

**Steps:** (a) write $H_0$, $H_1$ and the tail, (b) select the rows for the section, then find $n$, $\bar{x}$, $s$ and $df$, (c) compute $t$, (d) find the critical value and the p-value, (e) make a decision and write a conclusion, (f) check your answer with `stats.ttest_1samp(data, popmean, alternative=...)`.

*(15 marks)*

In [10]:
# ===================== Q6 – YOUR CODE HERE =====================
# t-test, two-tailed: column 'commute_min', Section C only, mu_0 = 35, sigma unknown

commute_C = np.array(df[df["section"] == "C"]["commute_min"])

x_bar = commute_C.mean()
Mu = 35
s = commute_C.std(ddof =1)
n = len(commute_C)
dof = n-1
t = (x_bar - Mu)/((s/np.sqrt(n)))
p_value = 2 * (1 - stats.t.cdf(abs(t), dof))

print("t-test:- ", t)
print("p-value:- ", p_value)

t-test:-  -2.86108144733173
p-value:-  0.014325046371812533


**Your answer – Q6**

- **$H_0$:The mean commute time of Section C students is 35 minutes**
- **$H_1$:The mean commute time of Section C students is different from 35 minutes**
- **Tail (left / right / two):two**
- **Test statistic:-2.861**
- **Critical value:±2.179**
- **p-value:0.0143**
- **Decision (Reject $H_0$ / Fail to reject $H_0$):Reject $H_0$**
- **Conclusion in words:Since the p-value is less than 0.05, we reject $H_0$. This means the average commute time of Section C students is significantly different from 35 minutes.**

---
## Q7. Summary Table *(10 marks)*

Fill in the table with your results, then answer the two short questions below it.

**Your answer – Q7**

| Q | Test | Tail | Statistic | Critical value | p-value | Decision |
|---|---|---|---|---|---|---|
| Q1 | Z-test | Left | -1.858 | -1.645 | 0.0316 | Reject $H_0$ |
| Q2 | Z-test | Right | 2.003 | 1.645 | 0.0226 | Reject $H_0$ |
| Q3 | Z-test | Two | -0.506 | ±1.960 | 0.613 | Fail to reject $H_0$ |
| Q4 | t-test | Left | -1.031 | -1.782 | 0.1615 | Fail to reject $H_0$ |
| Q5 | t-test | Right | 2.118 | 1.771 | 0.0269 | Reject $H_0$ |
| Q6 | t-test | Two | -2.861 | ±2.179 | 0.0143 | Reject $H_0$ |

**(a)** Why did Q1–Q3 use a Z-test, while Q4–Q6 used a t-test?

Because in Q1–Q3, the population standard deviation $\sigma$ was known, so a Z-test was appropriate. In Q4–Q6, $\sigma$ was not known and we used the sample standard deviation $s$, so a t-test was used instead.

**(b)** In Q3 and Q4 you failed to reject $H_0$. Does this prove that $H_0$ is true? Explain in one or two sentences.

No, failing to reject $H_0$ does not prove it is true. It only means that there is not enough evidence at the 0.05 significance level to reject it.

---
## Submission Guidelines

1. Complete every `YOUR CODE HERE` cell and every **Your answer** cell.
2. Before submitting, click **Kernel → Restart & Run All** (in Colab: **Runtime → Restart session and run all**). Make sure there are no errors and all outputs are visible.
3. Do not delete or change the starter cells.
4. Save the file as `RollNumber_Name_ZtTest.ipynb` (for example `23CS1042_Ananya_ZtTest.ipynb`).
5. Submit **only the `.ipynb` file** on the LMS before the deadline. PDF, HTML or `.py` files are not accepted.
6. This is an individual assignment. Do not copy code or answers from classmates.

**Marking:** Q1–Q6 are worth 15 marks each: hypotheses and tail (3), correct calculations (6), decision (3) and conclusion in words (3). Q7 is worth 10 marks. Total = 100.